File: reveal_password/reveal_password/page/security_dashboard/security_dashboard.py
# Copyright (c) 2025, Abhishek Chougule and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import now, add_to_date, getdate
import json


@frappe.whitelist()
def get_security_metrics(days=7):
	"""
	Get comprehensive security metrics for the dashboard.
	
	Args:
		days: Number of days to analyze
		
	Returns:
		Dictionary with all security metrics
	"""
	days = int(days)
	from_date = add_to_date(now(), days=-days)
	
	# Basic metrics
	total_sessions = frappe.db.count("Reveal Session", filters={"timestamp": [">=", from_date]})
	suspicious_count = frappe.db.count("Reveal Session", filters={
		"timestamp": [">=", from_date],
		"is_suspicious": 1
	})
	
	# Unique users
	unique_users = frappe.db.sql("""
		SELECT COUNT(DISTINCT user)
		FROM `tabReveal Session`
		WHERE timestamp >= %s
	""", (from_date,))[0][0]
	
	# Average anomaly score
	avg_anomaly = frappe.db.sql("""
		SELECT AVG(anomaly_score)
		FROM `tabReveal Session`
		WHERE timestamp >= %s AND anomaly_score > 0
	""", (from_date,))[0][0] or 0
	
	# Timeline data
	timeline_data = get_timeline_data(from_date, days)
	
	# Top users
	top_users = get_top_users(from_date)
	
	# Suspicious activities
	suspicious_activities = frappe.get_all(
		"Reveal Session",
		filters={
			"timestamp": [">=", from_date],
			"is_suspicious": 1
		},
		fields=["user", "doctype_revealed", "ip_address", "timestamp", 
		        "anomaly_score", "anomaly_reasons"],
		order_by="anomaly_score desc, timestamp desc",
		limit=20
	)
	
	# Top IPs
	top_ips = get_top_ips(from_date)
	
	# Device stats
	device_stats = get_device_stats(from_date, total_sessions)
	
	# Generate alerts
	alerts = generate_alerts(suspicious_count, total_sessions, avg_anomaly)
	
	return {
		"total_sessions": total_sessions,
		"suspicious_count": suspicious_count,
		"unique_users": unique_users,
		"avg_anomaly_score": avg_anomaly,
		"timeline_labels": timeline_data["labels"],
		"timeline_normal": timeline_data["normal"],
		"timeline_suspicious": timeline_data["suspicious"],
		"top_users_labels": top_users["labels"],
		"top_users_values": top_users["values"],
		"suspicious_activities": suspicious_activities,
		"top_ips": top_ips,
		"device_stats": device_stats,
		"alerts": alerts
	}


def get_timeline_data(from_date, days):
	"""Get timeline data for activity chart."""
	# Get daily counts
	data = frappe.db.sql("""
		SELECT 
			DATE(timestamp) as date,
			SUM(CASE WHEN is_suspicious = 0 THEN 1 ELSE 0 END) as normal,
			SUM(CASE WHEN is_suspicious = 1 THEN 1 ELSE 0 END) as suspicious
		FROM `tabReveal Session`
		WHERE timestamp >= %s
		GROUP BY DATE(timestamp)
		ORDER BY date ASC
	""", (from_date,), as_dict=True)
	
	# Fill in missing dates
	labels = []
	normal = []
	suspicious = []
	
	current_date = getdate(from_date)
	end_date = getdate(now())
	
	date_map = {str(d.date): d for d in data}
	
	while current_date <= end_date:
		date_str = str(current_date)
		labels.append(current_date.strftime("%d-%b"))
		
		if date_str in date_map:
			normal.append(date_map[date_str].normal)
			suspicious.append(date_map[date_str].suspicious)
		else:
			normal.append(0)
			suspicious.append(0)
		
		current_date = add_to_date(current_date, days=1)
	
	return {
		"labels": labels,
		"normal": normal,
		"suspicious": suspicious
	}


def get_top_users(from_date):
	"""Get top users by session count."""
	data = frappe.db.sql("""
		SELECT user, COUNT(*) as count
		FROM `tabReveal Session`
		WHERE timestamp >= %s
		GROUP BY user
		ORDER BY count DESC
		LIMIT 5
	""", (from_date,), as_dict=True)
	
	return {
		"labels": [d.user for d in data],
		"values": [d.count for d in data]
	}


def get_top_ips(from_date):
	"""Get top IP addresses with user information."""
	data = frappe.db.sql("""
		SELECT 
			ip_address,
			COUNT(*) as count,
			GROUP_CONCAT(DISTINCT user) as users
		FROM `tabReveal Session`
		WHERE timestamp >= %s AND ip_address IS NOT NULL
		GROUP BY ip_address
		ORDER BY count DESC
		LIMIT 10
	""", (from_date,), as_dict=True)
	
	for d in data:
		d.users = d.users.split(',') if d.users else []
	
	return data


def get_device_stats(from_date, total_sessions):
	"""Get device type statistics."""
	# Simple device detection from user agent
	data = frappe.db.sql("""
		SELECT user_agent, COUNT(*) as count
		FROM `tabReveal Session`
		WHERE timestamp >= %s AND user_agent IS NOT NULL
		GROUP BY user_agent
	""", (from_date,), as_dict=True)
	
	device_types = {
		"Desktop": 0,
		"Mobile": 0,
		"Tablet": 0,
		"Unknown": 0
	}
	
	for d in data:
		ua = d.user_agent.lower()
		if 'mobile' in ua or 'android' in ua or 'iphone' in ua:
			device_types["Mobile"] += d.count
		elif 'tablet' in ua or 'ipad' in ua:
			device_types["Tablet"] += d.count
		elif ua:
			device_types["Desktop"] += d.count
		else:
			device_types["Unknown"] += d.count
	
	stats = []
	for device_type, count in device_types.items():
		if count > 0:
			percentage = (count / total_sessions * 100) if total_sessions > 0 else 0
			stats.append({
				"device_type": device_type,
				"count": count,
				"percentage": round(percentage, 1)
			})
	
	return sorted(stats, key=lambda x: x["count"], reverse=True)


def generate_alerts(suspicious_count, total_sessions, avg_anomaly):
	"""Generate security alerts based on metrics."""
	alerts = []
	
	# High suspicious activity rate
	if total_sessions > 0:
		suspicious_rate = (suspicious_count / total_sessions) * 100
		
		if suspicious_rate > 20:
			alerts.append({
				"severity": "critical",
				"title": "Critical: High Suspicious Activity Rate",
				"message": f"{suspicious_rate:.1f}% of sessions are flagged as suspicious. Immediate investigation recommended."
			})
		elif suspicious_rate > 10:
			alerts.append({
				"severity": "warning",
				"title": "Warning: Elevated Suspicious Activity",
				"message": f"{suspicious_rate:.1f}% of sessions are flagged as suspicious. Monitor closely."
			})
	
	# High average anomaly score
	if avg_anomaly > 60:
		alerts.append({
			"severity": "critical",
			"title": "Critical: High Anomaly Score",
			"message": f"Average anomaly score is {avg_anomaly:.1f}/100. Review security policies."
		})
	elif avg_anomaly > 40:
		alerts.append({
			"severity": "warning",
			"title": "Warning: Elevated Anomaly Score",
			"message": f"Average anomaly score is {avg_anomaly:.1f}/100. Consider tightening security."
		})
	
	# Check for recent spike
	recent_suspicious = frappe.db.count("Reveal Session", filters={
		"timestamp": [">=", add_to_date(now(), hours=-24)],
		"is_suspicious": 1
	})
	
	if recent_suspicious > 10:
		alerts.append({
			"severity": "warning",
			"title": "Warning: Recent Activity Spike",
			"message": f"{recent_suspicious} suspicious activities in the last 24 hours."
		})
	
	# Info: All clear
	if not alerts:
		alerts.append({
			"severity": "info",
			"title": "All Clear",
			"message": "No significant security concerns detected."
		})
	
	return alerts


@frappe.whitelist()
def export_security_report(days=7):
	"""
	Export security report as CSV.
	
	Args:
		days: Number of days to include
		
	Returns:
		CSV string
	"""
	days = int(days)
	from_date = add_to_date(now(), days=-days)
	
	# Get all sessions
	sessions = frappe.get_all(
		"Reveal Session",
		filters={"timestamp": [">=", from_date]},
		fields=["user", "doctype_revealed", "document_name", "field_name",
		        "ip_address", "timestamp", "success", "is_suspicious",
		        "anomaly_score", "anomaly_reasons"],
		order_by="timestamp desc"
	)
	
	# Generate CSV
	import csv
	from io import StringIO
	
	output = StringIO()
	writer = csv.writer(output)
	
	# Header
	writer.writerow([
		"Timestamp", "User", "DocType", "Document", "Field",
		"IP Address", "Success", "Suspicious", "Anomaly Score", "Reasons"
	])
	
	# Data
	for session in sessions:
		writer.writerow([
			session.timestamp,
			session.user,
			session.doctype_revealed,
			session.document_name,
			session.field_name,
			session.ip_address or "N/A",
			"Yes" if session.success else "No",
			"Yes" if session.is_suspicious else "No",
			session.anomaly_score or 0,
			session.anomaly_reasons or "N/A"
		])
	
	return output.getvalue()


@frappe.whitelist()
def get_compliance_report(days=30):
	"""
	Generate compliance report for auditing.
	
	Args:
		days: Number of days to include
		
	Returns:
		Compliance report data
	"""
	days = int(days)
	from_date = add_to_date(now(), days=-days)
	
	# Total reveals
	total_reveals = frappe.db.count("Password Reveal Log", filters={
		"timestamp": [">=", from_date]
	})
	
	# Successful reveals
	successful_reveals = frappe.db.count("Password Reveal Log", filters={
		"timestamp": [">=", from_date],
		"success": 1
	})
	
	# Failed attempts
	failed_attempts = total_reveals - successful_reveals
	
	# Reveals by user
	by_user = frappe.db.sql("""
		SELECT user, COUNT(*) as count
		FROM `tabPassword Reveal Log`
		WHERE timestamp >= %s
		GROUP BY user
		ORDER BY count DESC
	""", (from_date,), as_dict=True)
	
	# Reveals by DocType
	by_doctype = frappe.db.sql("""
		SELECT revealed_doctype, COUNT(*) as count
		FROM `tabPassword Reveal Log`
		WHERE timestamp >= %s
		GROUP BY revealed_doctype
		ORDER BY count DESC
	""", (from_date,), as_dict=True)
	
	# MFA usage
	mfa_enabled_users = frappe.db.count("MFA Secret", filters={"is_enabled": 1})
	total_users = frappe.db.count("Trusted User", filters={"enabled": 1})
	
	return {
		"period": f"Last {days} days",
		"total_reveals": total_reveals,
		"successful_reveals": successful_reveals,
		"failed_attempts": failed_attempts,
		"success_rate": (successful_reveals / total_reveals * 100) if total_reveals > 0 else 0,
		"by_user": by_user,
		"by_doctype": by_doctype,
		"mfa_adoption": {
			"enabled": mfa_enabled_users,
			"total": total_users,
			"percentage": (mfa_enabled_users / total_users * 100) if total_users > 0 else 0
		}
	}

File: reveal_password/reveal_password/page/security_dashboard/security_dashboard.js
frappe.pages['security-dashboard'].on_page_load = function (wrapper) {
	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Security Dashboard',
		single_column: true
	});

	// Add action buttons
	page.set_primary_action('Export Report', () => export_security_report(), 'download');
	page.set_secondary_action('Refresh', () => load_dashboard(), 'refresh');

	// Add filter
	page.add_field({
		fieldname: 'time_range',
		label: __('Time Range'),
		fieldtype: 'Select',
		options: ['Last 24 Hours', 'Last 7 Days', 'Last 30 Days', 'Last 90 Days'],
		default: 'Last 7 Days',
		change: () => load_dashboard()
	});

	// Main container
	$(wrapper).find('.layout-main-section').append(`
		<div class="security-dashboard-container" style="padding: 20px;">
			<!-- Alert Section -->
			<div id="alerts-section" style="margin-bottom: 30px;"></div>
			
			<!-- Key Metrics -->
			<div class="row" style="margin-bottom: 30px;">
				<div class="col-md-3">
					<div class="metric-card" id="total-sessions">
						<div class="metric-icon"><i class="fa fa-chart-bar"></i></div>
						<div class="metric-value">0</div>
						<div class="metric-label">Total Sessions</div>
					</div>
				</div>
				<div class="col-md-3">
					<div class="metric-card" id="suspicious-count">
						<div class="metric-icon"><i class="fa fa-exclamation-triangle"></i></div>
						<div class="metric-value">0</div>
						<div class="metric-label">Suspicious Activities</div>
					</div>
				</div>
				<div class="col-md-3">
					<div class="metric-card" id="unique-users">
						<div class="metric-icon"><i class="fa fa-users"></i></div>
						<div class="metric-value">0</div>
						<div class="metric-label">Active Users</div>
					</div>
				</div>
				<div class="col-md-3">
					<div class="metric-card" id="avg-anomaly">
						<div class="metric-icon"><i class="fa fa-bullseye"></i></div>
						<div class="metric-value">0</div>
						<div class="metric-label">Avg Anomaly Score</div>
					</div>
				</div>
			</div>
			
			<!-- Charts Row -->
			<div class="row" style="margin-bottom: 30px;">
				<div class="col-md-8">
					<div class="dashboard-card">
						<h4>Activity Timeline</h4>
						<div id="activity-timeline-chart"></div>
					</div>
				</div>
				<div class="col-md-4">
					<div class="dashboard-card">
						<h4>Top Users</h4>
						<div id="top-users-chart"></div>
					</div>
				</div>
			</div>
			
			<!-- Suspicious Activity Table -->
			<div class="row" style="margin-bottom: 30px;">
				<div class="col-md-12">
					<div class="dashboard-card">
						<h4>Suspicious Activities</h4>
						<div id="suspicious-activities-table"></div>
					</div>
				</div>
			</div>
			
			<!-- IP Analysis -->
			<div class="row">
				<div class="col-md-6">
					<div class="dashboard-card">
						<h4>Top IP Addresses</h4>
						<div id="top-ips-list"></div>
					</div>
				</div>
				<div class="col-md-6">
					<div class="dashboard-card">
						<h4>Device Analysis</h4>
						<div id="device-analysis-list"></div>
					</div>
				</div>
			</div>
		</div>
	`);

	// Add custom CSS
	frappe.dom.set_style(`
		.metric-card {
			background: var(--card-bg);
			border: 1px solid var(--border-color);
			border-radius: 8px;
			padding: 20px;
			text-align: center;
			transition: transform 0.2s;
		}
		.metric-card:hover {
			transform: translateY(-2px);
			box-shadow: 0 4px 12px rgba(0,0,0,0.1);
		}
		.metric-icon {
			font-size: 32px;
			margin-bottom: 10px;
		}
		.metric-value {
			font-size: 28px;
			font-weight: 700;
			color: var(--text-color);
			margin-bottom: 5px;
		}
		.metric-label {
			font-size: 12px;
			color: var(--text-muted);
			text-transform: uppercase;
			letter-spacing: 0.5px;
		}
		.dashboard-card {
			background: var(--card-bg);
			border: 1px solid var(--border-color);
			border-radius: 8px;
			padding: 20px;
			min-height: 300px;
		}
		.alert-box {
			padding: 15px;
			border-radius: 8px;
			margin-bottom: 15px;
			border-left: 4px solid;
		}
		.alert-critical {
			background: #fee;
			border-color: #f44;
		}
		.alert-warning {
			background: #fffbeb;
			border-color: #f59e0b;
		}
		.alert-info {
			background: #eff6ff;
			border-color: #3b82f6;
		}
	`);

	function load_dashboard() {
		const time_range = page.fields_dict.time_range.get_value();
		const days = get_days_from_range(time_range);

		// Load security metrics
		frappe.call({
			method: 'reveal_password.reveal_password.page.security_dashboard.security_dashboard.get_security_metrics',
			args: { days: days },
			callback: function (r) {
				if (r.message) {
					render_dashboard(r.message);
				}
			}
		});
	}

	function get_days_from_range(range) {
		const map = {
			'Last 24 Hours': 1,
			'Last 7 Days': 7,
			'Last 30 Days': 30,
			'Last 90 Days': 90
		};
		return map[range] || 7;
	}

	function render_dashboard(data) {
		// Render alerts
		render_alerts(data.alerts);

		// Update metrics
		$('#total-sessions .metric-value').text(data.total_sessions);
		$('#suspicious-count .metric-value').text(data.suspicious_count);
		$('#unique-users .metric-value').text(data.unique_users);
		$('#avg-anomaly .metric-value').text(data.avg_anomaly_score.toFixed(1));

		// Render activity timeline
		new frappe.Chart("#activity-timeline-chart", {
			data: {
				labels: data.timeline_labels,
				datasets: [
					{
						name: "Normal",
						values: data.timeline_normal
					},
					{
						name: "Suspicious",
						values: data.timeline_suspicious
					}
				]
			},
			type: 'bar',
			height: 250,
			colors: ['#4CAF50', '#f44336']
		});

		// Render top users
		new frappe.Chart("#top-users-chart", {
			data: {
				labels: data.top_users_labels,
				datasets: [{
					values: data.top_users_values
				}]
			},
			type: 'pie',
			height: 250,
			colors: ['#3b82f6', '#8b5cf6', '#ec4899', '#f59e0b', '#10b981']
		});

		// Render suspicious activities table
		render_suspicious_table(data.suspicious_activities);

		// Render IP analysis
		render_ip_analysis(data.top_ips);

		// Render device analysis
		render_device_analysis(data.device_stats);
	}

	function render_alerts(alerts) {
		let html = '';

		alerts.forEach(alert => {
			const type_class = alert.severity === 'critical' ? 'alert-critical' :
				alert.severity === 'warning' ? 'alert-warning' : 'alert-info';
			html += `
				<div class="alert-box ${type_class}">
					<strong>${alert.title}</strong><br>
					${alert.message}
				</div>
			`;
		});

		$('#alerts-section').html(html || '<p style="color: #10b981;"><i class="fa fa-check"></i> No security alerts</p>');
	}

	function render_suspicious_table(activities) {
		let html = `<table class="table table-bordered table-hover">
			<thead>
				<tr>
					<th>User</th>
					<th>DocType</th>
					<th>IP Address</th>
					<th>Time</th>
					<th>Score</th>
					<th>Reasons</th>
				</tr>
			</thead>
			<tbody>`;

		activities.forEach(activity => {
			const score_color = activity.anomaly_score >= 75 ? 'red' :
				activity.anomaly_score >= 50 ? 'orange' : 'yellow';
			html += `
				<tr>
					<td>${activity.user}</td>
					<td>${activity.doctype_revealed}</td>
					<td><code>${activity.ip_address}</code></td>
					<td>${frappe.datetime.comment_when(activity.timestamp)}</td>
					<td><span class="indicator ${score_color}">${activity.anomaly_score}</span></td>
					<td style="font-size: 12px;">${activity.anomaly_reasons || 'N/A'}</td>
				</tr>
			`;
		});

		html += `</tbody></table>`;
		$('#suspicious-activities-table').html(html);
	}

	function render_ip_analysis(top_ips) {
		let html = '<div style="max-height: 300px; overflow-y: auto;">';

		top_ips.forEach(ip => {
			html += `
				<div style="padding: 10px; border-bottom: 1px solid var(--border-color);">
					<div style="display: flex; justify-content: space-between;">
						<code>${ip.ip_address}</code>
						<span class="badge badge-primary">${ip.count} sessions</span>
					</div>
					<div style="font-size: 12px; color: var(--text-muted); margin-top: 5px;">
						${ip.users.join(', ')}
					</div>
				</div>
			`;
		});

		html += '</div>';
		$('#top-ips-list').html(html);
	}

	function render_device_analysis(device_stats) {
		let html = '<div style="max-height: 300px; overflow-y: auto;">';

		device_stats.forEach(stat => {
			html += `
				<div style="padding: 10px; border-bottom: 1px solid var(--border-color);">
					<div style="display: flex; justify-content: space-between;">
						<strong>${stat.device_type}</strong>
						<span class="badge badge-secondary">${stat.count}</span>
					</div>
					<div style="font-size: 12px; color: var(--text-muted); margin-top: 5px;">
						${stat.percentage}% of total sessions
					</div>
				</div>
			`;
		});

		html += '</div>';
		$('#device-analysis-list').html(html);
	}

	function export_security_report() {
		const time_range = page.fields_dict.time_range.get_value();
		const days = get_days_from_range(time_range);

		frappe.call({
			method: 'reveal_password.reveal_password.page.security_dashboard.security_dashboard.export_security_report',
			args: { days: days },
			callback: function (r) {
				if (r.message) {
					// Download CSV
					const csv = r.message;
					const blob = new Blob([csv], { type: 'text/csv' });
					const url = window.URL.createObjectURL(blob);
					const a = document.createElement('a');
					a.href = url;
					a.download = `security_report_${frappe.datetime.now_date()}.csv`;
					a.click();

					frappe.show_alert({
						message: __('Report exported successfully'),
						indicator: 'green'
					}, 3);
				}
			}
		});
	}

	// Initialize
	load_dashboard();
}
